### PR TITLE
[Perf] Add MiniMax-M2 FP32 router GEMM kernel

### PR DESCRIPTION
**Pull request by qianlihuang**
## Purpose

This PR adds a specialized FP32 router GEMM path for MiniMax-M2. This is a scoped version of the MiniMax-M2 gate optimization from #38445 without the NVFP4 path. Credit to @jeejeelee for the original direction and implementation work.

related discussion https://vllm-dev.slack.com/archives/C09D4R4MYLR/p1775591605037309

## Test Plan

<details>
<summary>Correctness and benchmark commands</summary>

Correctness:

```bash
CUDA_VISIBLE_DEVICES=0 pytest -q -s tests/kernels/test_fp32_router_gemm.py
```

Additional max-error check against `torch.nn.functional.linear(x.float(), w)`:

```bash
CUDA_VISIBLE_DEVICES=0 python3 - <<'PY'
import torch
import torch.nn.functional as F
from vllm import _custom_ops as ops

torch.manual_seed(0)
H, E = 3072, 256

for precision, matmul_precision in [("default", None), ("fp32_high", "high")]:
    if matmul_precision is not None:
        torch.set_float32_matmul_precision(matmul_precision)
    else:
        torch.set_float32_matmul_precision("highest")

    for dtype, atol in [(torch.bfloat16, 2e-2), (torch.float32, 2e-4)]:
        for M in [1, 2, 3, 4, 7, 8, 15, 16, 31, 32]:
            x = torch.randn(M, H, device="cuda", dtype=dtype).contiguous()
            w = torch.randn(E, H, device="cuda", dtype=torch.float32).contiguous()
            out = ops.fp32_router_gemm(x, w)
            ref = F.linear(x.float(), w)
            err = (out - ref).abs()
            print(f"prec={precision} {dtype} M={M}",
                  "max_abs=", err.max().item(),
                  "mean_abs=", err.mean().item(),
                  "pass=", bool(torch.allclose(out, ref, atol=atol, rtol=0)))
PY
```

Accuracy:

```bash
python3 tests/evals/gsm8k/gsm8k_eval.py
```

Microbenchmark:

```bash
CUDA_VISIBLE_DEVICES=0 TRTLLM_ENABLE_PDL=0 \
python3 benchmarks/kernels/benchmark_router_gemm.py \
  --model MiniMaxAI/MiniMax-M2.7 \
  --max-batch-size 32 \
  --trust-remote-code

CUDA_VISIBLE_DEVICES=0 TRTLLM_ENABLE_PDL=1 \
python3 benchmarks/kernels/benchmark_router_gemm.py \
  --model MiniMaxAI/MiniMax-M2.7 \
  --max-batch-size 32 \
  --trust-remote-code
```

Serving benchmark:

Server launch variants tested:

```bash
TRTLLM_ENABLE_PDL=1 vllm serve /model/minimax-m2.7 \
  -tp 4 \
  --tool-call-parser minimax_m2 \
  --reasoning-parser minimax_m2 \
  --enable-auto-tool-choice \
  --trust-remote-code \
  --no-enable-prefix-caching

TRTLLM_ENABLE_PDL=0 vllm serve /model/minimax-m2.7 \
  -tp 4 \
  --tool-call-parser minimax_m2 \
  --reasoning-parser minimax_m2 \
  --enable-auto-tool-choice \
  --trust-remote-code \
  --no-enable-prefix-caching
```

Benchmark command:

```bash
export MODEL=/model/minimax-m2.7
export PORT=8000

vllm bench serve \
  --backend openai-chat \
  --model "$MODEL" \
  --endpoint /v1/chat/completions \
  --host 127.0.0.1 \
  --port "$PORT" \
  --dataset-name random \
  --num-prompts 128 \
  --random-input-len 8192 \
  --random-output-len 1024 \
  --random-range-ratio 0.0 \
  --num-warmups 16 \
  --request-rate inf \
  --max-concurrency 16 \
  --save-result

vllm bench serve \
  --backend openai-chat \
  --model "$MODEL" \
  --endpoint /v1/chat/completions \
  --host 127.0.0.1 \
  --port "$PORT" \
  --dataset-name random \
  --num-prompts 128 \
  --random-input-len 8192 \
  --random-output-len 1024 \
  --random-range-ratio 0.0 \
  --num-warmups 16 \
  --request-rate inf \
  --max-concurrency 8 \
  --save-result
```

</details>

## Test Result

Summary: serving benchmark on H800, TP=4

| Config | Concurrency | Total tok/s | Mean TPOT (ms) | Total tok/s delta | TPOT delta |
|---|---:|---:|---:|---:|---:|
| before PR | 16 | 7790.78 | 17.18 | baseline | baseline |
| after PR, `TRTLLM_ENABLE_PDL=0` | 16 | 8239.03 | 16.02 | +5.8% | -6.8% |
| after PR, `TRTLLM_ENABLE_PDL=1` | 16 | 8127.72 | 16.30 | +4.3% | -5.1% |
| before PR | 8 | TBD | TBD | baseline | baseline |
| after PR, `TRTLLM_ENABLE_PDL=0` | 8 | 5742.62 | 11.40 | TBD | TBD |
| after PR, `TRTLLM_ENABLE_PDL=1` | 8 | 5722.89 | 11.46 | TBD | TBD |

Accuracy:

| Metric | Before | After (`TP4`, `PDL=0/1`) |
|---|---:|---:|
| gsm8k | ~0.92 | 0.920 / 0.923 |

<details>
<summary>Correctness results</summary>

Correctness:

```text
13 passed, 16 warnings in 3.26s
```

Additional max-error check:

```text
prec=default torch.bfloat16 M=1 max_abs= 3.814697265625e-05 mean_abs= 9.896233677864075e-06 pass= True
prec=default torch.bfloat16 M=2 max_abs= 6.103515625e-05 mean_abs= 1.0900897905230522e-05 pass= True
prec=default torch.bfloat16 M=3 max_abs= 4.673004150390625e-05 mean_abs= 1.0053938240162097e-05 pass= True
prec=default torch.bfloat16 M=4 max_abs= 3.814697265625e-05 mean_abs= 1.0253861546516418e-05 pass= True
prec=default torch.bfloat16 M=7 max_abs= 6.103515625e-05 mean_abs= 1.2296851309656631e-05 pass= True
prec=default torch.bfloat16 M=8 max_abs= 5.340576171875e-05 mean_abs= 1.2036529369652271e-05 pass= True
prec=default torch.bfloat16 M=15 max_abs= 5.53131103515625e-05 mean_abs= 1.2322170732659288e-05 pass= True
prec=default torch.bfloat16 M=16 max_abs= 6.103515625e-05 mean_abs= 1.2327080185059458e-05 pass= True
prec=default torch.bfloat16 M=31 max_abs= 6.103515625e-05 mean_abs= 9.407965080754366e-06 pass= True
prec=default torch.bfloat16 M=32 max_abs= 6.103515625e-05 mean_abs= 9.411050996277481e-06 pass= True
prec=default torch.float32 M=1 max_abs= 3.0517578125e-05 mean_abs= 8.818693459033966e-06 pass= True
prec=default torch.float32 M=2 max_abs= 4.57763671875e-05 mean_abs= 9.83034260571003e-06 pass= True
prec=default torch.float32 M=3 max_abs= 4.57763671875e-05 mean_abs= 9.48490014707204e-06 pass= True
prec=default torch.float32 M=4 max_abs= 5.340576171875e-05 mean_abs= 9.826384484767914e-06 pass= True
prec=default torch.float32 M=7 max_abs= 5.7220458984375e-05 mean_abs= 1.2004881682514679e-05 pass= True
prec=default torch.float32 M=8 max_abs= 5.340576171875e-05 mean_abs= 1.2546559446491301e-05 pass= True
prec=default torch.float32 M=15 max_abs= 6.103515625e-05 mean_abs= 1.2333870472502895e-05 pass= True
prec=default torch.float32 M=16 max_abs= 7.62939453125e-05 mean_abs= 1.221662387251854e-05 pass= True
prec=default torch.float32 M=31 max_abs= 7.62939453125e-05 mean_abs= 9.509156370768324e-06 pass= True
prec=default torch.float32 M=32 max_abs= 7.62939453125e-05 mean_abs= 9.350329492008314e-06 pass= True
prec=fp32_high torch.bfloat16 M=1 max_abs= 3.0517578125e-05 mean_abs= 8.903443813323975e-06 pass= True
prec=fp32_high torch.bfloat16 M=2 max_abs= 0.044440269470214844 mean_abs= 0.009015086106956005 pass= False
prec=fp32_high torch.bfloat16 M=3 max_abs= 0.0346832275390625 mean_abs= 0.008942199870944023 pass= False
prec=fp32_high torch.bfloat16 M=4 max_abs= 0.04113006591796875 mean_abs= 0.009151562117040157 pass= False
prec=fp32_high torch.bfloat16 M=7 max_abs= 0.04279804229736328 mean_abs= 0.008998721837997437 pass= False
prec=fp32_high torch.bfloat16 M=8 max_abs= 0.0445556640625 mean_abs= 0.009134004823863506 pass= False
prec=fp32_high torch.bfloat16 M=15 max_abs= 0.03853607177734375 mean_abs= 0.009143132716417313 pass= False
prec=fp32_high torch.bfloat16 M=16 max_abs= 0.04340171813964844 mean_abs= 0.00932237133383751 pass= False
prec=fp32_high torch.bfloat16 M=31 max_abs= 0.04761219024658203 mean_abs= 0.009131493046879768 pass= False
prec=fp32_high torch.bfloat16 M=32 max_abs= 0.04586029052734375 mean_abs= 0.009119704365730286 pass= False
prec=fp32_high torch.float32 M=1 max_abs= 3.4332275390625e-05 mean_abs= 8.806586265563965e-06 pass= True
prec=fp32_high torch.float32 M=2 max_abs= 0.05176544189453125 mean_abs= 0.013077953830361366 pass= False
prec=fp32_high torch.float32 M=3 max_abs= 0.051513671875 mean_abs= 0.01334131509065628 pass= False
prec=fp32_high torch.float32 M=4 max_abs= 0.05275726318359375 mean_abs= 0.013239540159702301 pass= False
prec=fp32_high torch.float32 M=7 max_abs= 0.06198310852050781 mean_abs= 0.013133209198713303 pass= False
prec=fp32_high torch.float32 M=8 max_abs= 0.06512451171875 mean_abs= 0.013000346720218658 pass= False
prec=fp32_high torch.float32 M=15 max_abs= 0.06135368347167969 mean_abs= 0.013118076138198376 pass= False
prec=fp32_high torch.float32 M=16 max_abs= 0.061614990234375 mean_abs= 0.012958699837327003 pass= False
prec=fp32_high torch.float32 M=31 max_abs= 0.06546783447265625 mean_abs= 0.012923624366521835 pass= False
prec=fp32_high torch.float32 M=32 max_abs= 0.07131576538085938 mean_abs= 0.012893503531813622 pass= False
```

</details>

<details>
<summary>Accuracy results</summary>

Command:

```bash
python3 tests/evals/gsm8k/gsm8k_eval.py
```

| Metric | Before | After (`TP4`, `PDL=0`) | After (`TP4`, `PDL=1`) |
|---|---:|---:|---:|
| gsm8k | ~0.92 | 0.920 | 0.923 |

</details>

<details>
<summary>Router GEMM microbenchmark results</summary>

Microbenchmark on Hopper, `TRTLLM_ENABLE_PDL=0`:

| batch_size | PyTorch TFLOPs | vLLM TFLOPs |
|---:|---:|---:|
| 1 | 0.312044 | 0.776409 |
| 2 | 0.267624 | 1.491871 |
| 4 | 0.525618 | 2.600903 |
| 8 | 0.787964 | 4.359729 |
| 16 | 1.515117 | 6.503215 |
| 32 | 3.487280 | 7.970897 |

Microbenchmark on Hopper, `TRTLLM_ENABLE_PDL=1`:

| batch_size | PyTorch TFLOPs | vLLM TFLOPs |
|---:|---:|---:|
| 1 | 0.310701 | 0.916735 |
| 2 | 0.263064 | 1.719504 |
| 4 | 0.522884 | 3.029809 |
| 8 | 0.783282 | 4.936780 |
| 16 | 1.514976 | 6.998538 |
| 32 | 3.495466 | 8.353478 |

</details>

<details>
<summary>Full serving benchmark outputs</summary>

TP4 before PR:

```text
============ Serving Benchmark Result ============
Successful requests:                     128
Failed requests:                         0
Maximum request concurrency:             16
Benchmark duration (s):                  152.04
Total input tokens:                      1053440
Total generated tokens:                  131072
Request throughput (req/s):              0.84
Output token throughput (tok/s):         862.09
Peak output token throughput (tok/s):    1039.00
Peak concurrent requests:                26.00
Total token throughput (tok/s):          7790.78
---------------Time to First Token----------------
Mean TTFT (ms):                          1417.27
Median TTFT (ms):                        1432.74
P99 TTFT (ms):                           4043.75
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          17.18
Median TPOT (ms):                        17.21
P99 TPOT (ms):                           18.69
---------------Inter-token Latency----------------
Mean ITL (ms):                           22.76
Median ITL (ms):                         15.93
P99 ITL (ms):                            235.74
==================================================
```

TP4 after PR, `TRTLLM_ENABLE_PDL=0`:

```text
============ Serving Benchmark Result ============
Successful requests:                     128
Failed requests:                         0
Maximum request concurrency:             16
Benchmark duration (s):                  143.77
Total input tokens:                      1053440
Total generated tokens:                  131072
Request throughput (req/s):              0.89
Output token throughput (tok/s):         911.69
Peak output token throughput (tok/s):    1104.00
Peak concurrent requests:                26.00
Total token throughput (tok/s):          8239.03
---------------Time to First Token----------------
Mean TTFT (ms):                          1574.59
Median TTFT (ms):                        1644.48
P99 TTFT (ms):                           4183.01
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          16.02
Median TPOT (ms):                        16.11
P99 TPOT (ms):                           17.38
---------------Inter-token Latency----------------
Mean ITL (ms):                           21.43
Median ITL (ms):                         14.99
P99 ITL (ms):                            233.26
==================================================
============ Serving Benchmark Result ============
Successful requests:                     128
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  206.27
Total input tokens:                      1053440
Total generated tokens:                  131072
Request throughput (req/s):              0.62
Output token throughput (tok/s):         635.45
Peak output token throughput (tok/s):    736.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          5742.62
---------------Time to First Token----------------
Mean TTFT (ms):                          1222.23
Median TTFT (ms):                        1407.24
P99 TTFT (ms):                           2039.90
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.40
Median TPOT (ms):                        11.33
P99 TPOT (ms):                           12.45
---------------Inter-token Latency----------------
Mean ITL (ms):                           15.35
Median ITL (ms):                         11.15
P99 ITL (ms):                            22.48
==================================================
```

TP4 after PR, `TRTLLM_ENABLE_PDL=1`:

```text
============ Serving Benchmark Result ============
Successful requests:                     128
Failed requests:                         0
Maximum request concurrency:             16
Benchmark duration (s):                  145.74
Total input tokens:                      1053440
Total generated tokens:                  131072
Request throughput (req/s):              0.88
Output token throughput (tok/s):         899.37
Peak output token throughput (tok/s):    1104.00
Peak concurrent requests:                25.00
Total token throughput (tok/s):          8127.72
---------------Time to First Token----------------
Mean TTFT (ms):                          1532.83
Median TTFT (ms):                        1434.39
P99 TTFT (ms):                           4178.46
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          16.30
Median TPOT (ms):                        16.38
P99 TPOT (ms):                           17.64
---------------Inter-token Latency----------------
Mean ITL (ms):                           21.26
Median ITL (ms):                         15.08
P99 ITL (ms):                            233.00
==================================================
============ Serving Benchmark Result ============
Successful requests:                     128
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  206.98
Total input tokens:                      1053440
Total generated tokens:                  131072
Request throughput (req/s):              0.62
Output token throughput (tok/s):         633.27
Peak output token throughput (tok/s):    744.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          5722.89
---------------Time to First Token----------------
Mean TTFT (ms):                          1211.84
Median TTFT (ms):                        1412.39
P99 TTFT (ms):                           2036.89
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.46
Median TPOT (ms):                        11.37
P99 TPOT (ms):                           12.47
---------------Inter-token Latency----------------
Mean ITL (ms):                           15.29
Median ITL (ms):                         11.15
P99 ITL (ms):                            22.51
==================================================
```

</details>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by chaunceyjiang** [comment]
hi  @jeejeelee I’d like to hear your thoughts on this.
